webrender/src/render_backend.rs

```diff
@@ -310,7 +318,21 @@ impl RenderBackend {
         let mut new_pipeline_sizes = HashMap::new();
 
         if let Some(id) = self.current_bound_webgl_context_id {
-            self.webgl_contexts.get(&id).unwrap().unbind().unwrap();
+            self.webgl_contexts.get(&id).unwrap().unbind();
+            self.current_bound_webgl_context_id = None;
+        }
+
+        // When running in OSMesa mode with texture sharing,
+        // a flush is required on any GL contexts to ensure
+        // that read-back from the shared texture returns
+        // valid data! This should be fine to have run on all
+        // implementations - a single flush for each webgl
+        // context at the start of a render frame should
+        // incur minimal cost.
+        for (_, webgl_context) in &self.webgl_contexts {
+            webgl_context.make_current();
+            gl::flush();
```

Comment:
I'm fine with this, but are you sure the `if let Some(id)` before this loop isn't a better place to do it? I expect a context change to trigger a flush. Indeed, from a quick look at mesa's source:

```c
   if (curCtx && 
       (curCtx->WinSysDrawBuffer || curCtx->WinSysReadBuffer) &&
       /* make sure this context is valid for flushing */
       curCtx != newCtx &&
       curCtx->Const.ContextReleaseBehavior ==
       GL_CONTEXT_RELEASE_BEHAVIOR_FLUSH)
      _mesa_flush(curCtx);
```

It seems to be the case (the `WinSysDrawBuffer` and `WinSysReadBuffer` are set on `make_current`, and `GL_CONTEXT_RELEASE_BEHAVIOR_FLUSH` is the default one (and we're not changing it anywhere). So the problem is we're not flushing on `unbind` on the OSMesa back-end.

I'm sending a patch to allow unbinding a context in OSMesa to the mesa-dev mailing list, meanwhile I think the cleanest solution is to call `gl::flush` in `OSMesaContext::unbind`, what do you think?